### PR TITLE
fix(desktop): present microphone alerts without blocking

### PR DESCRIPTION
**Pull request by undivisible**
## Summary

Present microphone recovery alerts as non-blocking sheets instead of blocking the main actor. When no Omi window is available, queue the alert until the application becomes active or a window becomes key.

## Sentry issue

- [OMI-DESKTOP-3M5](https://omi-nk3.sentry.io/issues/7642270321/?project=4511086024851456)

Related: #11421 prevents repeated silent-mic recovery from reaching this alert; this PR removes the separate blocking modal when an alert is presented.

## Failure class

Failure-Class: none

## Verification

- Swift format, parse, test-quality, e2e coverage, and diff checks passed.
- The focused XCTest compiled and linked locally; its runtime is blocked by the host's missing Sparkle.framework.

---

> [!NOTE]
> **Medium Risk**
> Changes main-thread alert presentation and window activation for all `showAlert` call sites (transcription, permissions, device errors); incorrect queueing or summon logic could hide alerts or steal focus, though tests cover the main edge cases.
> 
> **Overview**
> Replaces **blocking** `NSAlert.runModal()` for shared `showAlert` paths with a **non-blocking sheet** on the main shell window, so microphone and transcription warnings no longer freeze the main actor (addresses the silent-mic / permission alert freeze).
> 
> Introduces **`DesktopAlertPresenting`** and **`AppKitSheetAlertPresenter`**, wired via `AppState.alertPresenter`. The presenter **queues** alerts, **coalesces** duplicates, **reveals** the main window when no shell is presentable, **defers** when another sheet is attached, and runs optional **completions after dismiss**—so opening **System Settings → Microphone** happens only after the user dismisses **"Omi Needs Microphone Access"**, not underneath it.
> 
> Adds **`AppStateAlertPresentationTests`** and extends the audio-recording e2e flow for silent-mic recovery and permission alert behavior; changelog notes the freeze fix.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 24ecd0cf8a5d24d3ebc4744ff90db63f58b69cbd. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

Line-Count-Exception: desktop/macos/Desktop/Sources/AppState/AppState+Transcription.swift | 1738 -> 1743 | permission alert gains a sheet-completion callback that opens the microphone settings pane only after the non-blocking alert is seen

### CODE REVIEW

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for taking on the blocking microphone alert path. I agree with the direction of moving `showAlert` away from `NSAlert.runModal()`, but I think this needs one production fix before merge.

Blocking issue:
- `desktop/macos/Desktop/Sources/AppState/AppState+Permissions.swift`: `AppKitSheetAlertPresenter.present` now returns after only logging when `NSApp.keyWindow`, `NSApp.mainWindow`, and visible windows are all unavailable. That is a behavior regression from the old `runModal()` path: the silent-microphone warning, `Device Not Connected`, and `Couldn't Start Transcription` can now be dropped entirely in menu-bar / no-visible-window states. Please preserve a user-visible fallback, or queue/defer the non-blocking sheet until an appropriate app window exists, rather than losing the alert.

Other file-specific notes:
- `desktop/macos/Desktop/Sources/AppState.swift`: the `DesktopAlertPresenting` seam and `alertPresenter` injection are a good way to make `showAlert` testable, but this also makes the default presenter behavior the critical production path to cover.
- `desktop/macos/Desktop/Tests/AppStateAlertPresentationTests.swift`: the new XCTest verifies delegation through the seam, but it stubs out `AppKitSheetAlertPresenter`, so it would still pass if the real presenter drops alerts in the no-window case or uses the wrong presentation mode. Please add coverage for the real presenter/window-selection behavior, or extract the AppKit window/alert operations behind a smaller seam that can prove `beginSheetModal` and fallback behavior.
- `desktop/macos/changelog/unreleased/20260812-silent-microphone-alert.json`: the changelog entry matches the intended user-facing fix, but it should only land once alerts are guaranteed not to disappear when no sheet host is available.
- `desktop/macos/e2e/flows/audio-recording.yaml`: the new S4B flow is the right manual scenario; please make sure it explicitly covers a no/hidden-main-window state if that is a supported desktop usage mode.

CI note: current Desktop Swift CI is not green (`Desktop Swift Static & Test Contracts` was cancelled and `Desktop Swift Build & Tests` failed because the required verification result was cancelled). I’m requesting changes for the production alert-dropping path above; CI should also be rerun once that is fixed.

Leaving this for the macOS desktop alert behavior to be fixed and verified before merge.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by undivisible** [comment]
Addressed the remaining open threads from the 2026-08-12 codex review on head `d24eb2fe7a` (pushed `31a1f65753`):

- **P2 — defer while the shell already owns a sheet**: `AppKitSheetAlertPresenter` now checks `canHostSheet` (`window.attachedSheet == nil`) before presenting; when the shell hosts another sheet (e.g. a SwiftUI sheet), the request stays queued and presentation retries on `NSWindow.didEndSheetNotification`. A second `beginSheetModal` can no longer raise on a sheet-owning window.
- **P1 — reveal ordered-in shells before presenting**: `shellWindowProvider` now requires `NSApp.isActive` (via `presentableShellWindow(_:isActive:)`), mirroring `ShellSummon.toggleAction`. A visible-but-inactive shell routes through `openMainAppWindow` / `ShellSummon.summon` before the sheet attaches, so the warning lands in front.

New regression coverage exercises the real presenter through injected seams: deferral-until-sheet-ends drains queued alerts in order, reveal-when-no-window-remains, and the active-app gate. Ran `./scripts/dev-feedback.py --once swift 'AppStateAlertPresentationTests'` → 8/8 passed (loaded Sparkle via `DYLD_FRAMEWORK_PATH` against the signed bundle's copy — the known local test-bundle dylib gap). `scripts/check_desktop_test_quality.py` green.

**Review by kodjima33 — APPROVED** [review]
Non-blocking presentation with a queue until a window is key is the right shape, and it pairs with #11421 rather than duplicating it. Held from merge only by the stale desktop Swift red (fixed on main by #11504; the PR's merge ref predates it).

File: desktop/macos/Desktop/Sources/AppState/AppState+Permissions.swift [hunk 1]
```diff
@@ -11,6 +11,79 @@ import SwiftUI
...
+      object: nil)
+  }
+
+  deinit {
+    NotificationCenter.default.removeObserver(self)
+  }
+
+  func present(title: String, message: String) {
+    pendingAlerts.append(.init(title: title, message: message))
+    presentPendingAlertIfPossible()
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/AppState/AppState+Permissions.swift]: Deduplication is implemented: `present(title:message:)` coalesces via `guard activeAlert != request, !pendingAlerts.contains(request)` (AppState+Permissions.swift:85) before enqueueing, so repeated identical requests while one is active or pending cannot queue duplicates. Covered by `testAppKitPresenterCoalescesMatchingPendingAlerts` and `testAppKitPresenterCoalescesMatchingActiveAlerts`.

File: desktop/macos/Desktop/Tests/AppStateAlertPresentationTests.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+@testable import Omi_Computer
+
+@MainActor
+final class AppStateAlertPresentationTests: XCTestCase {
+  func testShowAlertPresentsThroughTheNonBlockingSheetSeam() {
+    let state = AppState()
+    let presenter = RecordingAlertPresenter()
+    state.alertPresenter = presenter
+
+    state.showAlert(title: "Microphone Isn't Capturing Audio", message: "Check your input device and try again.")
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Tests/AppStateAlertPresentationTests.swift]: Addressed in current head: `testShowAlertPresentsThroughTheNonBlockingSheetSeam` now exercises the production `AppKitSheetAlertPresenter` with an injected `appKitOperations` recorder (asserting the sheet operation is what production selects), replacing the stub presenter. The other tests in this file already exercise the real presenter for reveal/defer/coalesce paths.

**Comment by undivisible** [comment]
@Git-on-my-level Addressed the P2 'delay opening Settings until the alert completes' on head `98504c69e1` (pushed just now).

`DesktopAlertPresenting.present` now accepts an optional per-alert `completion`, and `AppKitSheetAlertPresenter` runs it when that alert's sheet ends (not when the request is enqueued). `surfaceMicrophonePermissionAlert` therefore defers `NSWorkspace.open` (Microphone pane) until after the non-blocking permission sheet has been seen — the old `runModal()` path implicitly blocked here, so the pane now opens at the same logical point.

Coverage: `testAppKitPresenterRunsCompletionAfterTheSheetEnds` proves the completion runs after the sheet ends, not before (using a deferred-completion recorder). Verified locally: `xcrun swift test --filter AppStateAlertPresentationTests` = 9/9 passing, swift-format + swiftlint + test-quality ratchet clean. CI should re-run on the new head.

**Comment by undivisible** [comment]
## Re-review request — blocker resolved (head `391d25cc05`)

@Git-on-my-level the no-window/drop blocker and the remaining P2 thread are addressed. Commits `97747c4c5b`–`98504c69` already landed the defer-until-window design (queue + reveal + coalescing + active-sheet gate) with tests exercising the real `AppKitSheetAlertPresenter` via the injected `AppKitAlertOperations`/`shellWindowProvider` seams; `commit 020ebf769a` (this round) closes the last open thread:

- **Pause queue draining after a completion that drops the foreground.** After any sheet's completion runs, the presenter only continues draining while Omi is still the active app. When the microphone-permission alert's completion opens System Settings (deactivating Omi), the next queued alert is held rather than summoning Omi back over the settings pane; the existing `didBecomeActive`/`didBecomeKey`/`didEndSheet` observers resume the queue when the user returns.
- **Regression test:** `testAppKitPresenterPausesQueuedAlertsAfterACompletionDropsTheForeground` — ends the permission sheet with Omi deactivated and asserts the queued `Device Not Connected` alert is NOT dragged forward, then presents once Omi is active again.

**Verification (local):** `AppStateAlertPresentationTests` ran 10/10 green (`swift test --filter AppStateAlertPresentationTests`). Local toolchain caveat: this repo's CI pins Xcode 16.4; the machine only has Xcode 27 (Swift 6.4), which rejects an unrelated pre-existing file (`ShellClickThrough.swift`, newer `ImplicitStrongCapture` diagnostic — root-caused in main issue #11563/#11573), so a temporary local-only, un-reverted build workaround was applied just long enough to run the suite, then reverted. Ratchet gates pass: `check_desktop_test_quality.py` green, `swift-format` lint green, `product-file-line-count-ratchet` OK after merging `origin/main`.

**CI status — pre-existing main red, not this PR:** `Desktop Swift Release Compile`, `Static & Test Contracts`, and `Build & Tests` were already red at the untouched head `98504c69` (before this round's commit) and remain red on `main` itself — tracked in open issues #11563 (77 failures/3 suites on main) and #11573 (3 suites SIGSEGV on CI only), root-caused to `15a51fb985`. The current failures are `signal 11` segfaults in the release `-O`/`emit-module` compile of the Omi_Computer module (which abort before any test runs) plus an unrelated `desktop launcher script` harness failure — none exercise the alert-presenter change, and the alert suite passes locally under a stricter toolchain.

**Comment by undivisible** [comment]
Addressed the post-rebase Codex P2 on head (pushed just now): pause draining before launching System Settings.

`DesktopAlertPresenting.pauseQueueUntilAppActive()` is an explicit request. The microphone-permission completion calls it *before* `NSWorkspace.open`, because that hand-off can return while Omi is still the active app. The presenter then refuses to drain until `NSApplication.didBecomeActive`. Inferring the pause from `shellWindowProvider()` synchronously is no longer the gate.

New coverage: `testAppKitPresenterHonorsAnExplicitQueuePauseFromCompletion` keeps the shell window presentable through the completion and asserts the queued alert stays queued until `didBecomeActive`.

**Review by kodjima33 — APPROVED** [review]
Desktop mic-alert bug fix — re-confirmed still approve-only: Desktop Swift Build/Release/Contracts CI and Backend Hermetic Merge Gate are still failing on this head, and reviewDecision remains CHANGES_REQUESTED from a human. Not overriding either; approving to keep pace with new commits, no merge.
